File: dms/api/inventory.py
# Copyright (c) 2026, Mania and contributors

import frappe

from dms.dealer_management_system.utils.inventory_dashboard import (
	get_inventory_dashboard_defaults,
	get_inventory_insights,
	get_spare_part_stock_balance,
	get_spare_part_stock_ledger,
)


@frappe.whitelist()
def get_inventory_defaults(company=None):
	frappe.has_permission("Stock Entry", "read", throw=True)
	return get_inventory_dashboard_defaults(company)


@frappe.whitelist()
def get_stock_balance_report(
	company=None,
	warehouse=None,
	item_code=None,
	item_group=None,
	search=None,
	as_on_date=None,
	sort_order="desc",
	limit=500,
):
	return get_spare_part_stock_balance(
		company=company,
		warehouse=warehouse,
		item_code=item_code,
		item_group=item_group,
		search=search,
		as_on_date=as_on_date,
		sort_order=sort_order,
		limit=int(limit or 500),
	)


@frappe.whitelist()
def get_stock_ledger_report(
	company=None,
	warehouse=None,
	item_code=None,
	item_group=None,
	search=None,
	from_date=None,
	to_date=None,
	limit=200,
):
	return get_spare_part_stock_ledger(
		company=company,
		warehouse=warehouse,
		item_code=item_code,
		item_group=item_group,
		search=search,
		from_date=from_date,
		to_date=to_date,
		limit=int(limit or 200),
	)


@frappe.whitelist()
def get_inventory_insights_report(
	company=None,
	warehouse=None,
	from_date=None,
	to_date=None,
	low_stock_limit=25,
	consumed_limit=25,
):
	return get_inventory_insights(
		company=company,
		warehouse=warehouse,
		from_date=from_date,
		to_date=to_date,
		low_stock_limit=int(low_stock_limit or 25),
		consumed_limit=int(consumed_limit or 25),
	)

File: frontend/services/inventory.ts
/**
 * Inventory dashboard — dms.api.inventory
 */
import { apiRequest } from './apiClient';
import type { DmsWarehouseOption } from './stockOperations';

const API = 'dms.api.inventory';

export interface InventoryDefaults {
  company: string;
  default_warehouse?: string | null;
  warehouses: DmsWarehouseOption[];
  companies: string[];
  item_groups: string[];
  as_on_date: string;
}

export interface StockBalanceRow {
  spare_part: string;
  item_code: string;
  item_name: string;
  item_group?: string;
  stock_uom?: string;
  oem_part_number?: string;
  bin_location?: string;
  minimum_stock_level: number;
  qty: number;
  is_low_stock: boolean;
  warehouse?: string | null;
}

export interface StockBalanceReport {
  rows: StockBalanceRow[];
  summary: {
    item_count: number;
    total_qty: number;
    low_stock_count: number;
    warehouse_count: number;
    as_on_date: string;
  };
}

export interface StockLedgerRow {
  posting_date: string;
  posting_time?: string;
  item_code: string;
  item_name: string;
  warehouse?: string;
  actual_qty: number;
  qty_after_transaction: number;
  voucher_type?: string;
  voucher_no?: string;
  stock_uom?: string;
}

export interface StockLedgerReport {
  rows: StockLedgerRow[];
  summary: {
    from_date: string;
    to_date: string;
    entry_count: number;
  };
}

export interface InventoryInsightsReport {
  low_stock: StockBalanceRow[];
  most_consumed: { item_code: string; item_name: string; consumed_qty: number }[];
  summary: {
    low_stock_count: number;
    from_date: string;
    to_date: string;
  };
}

export async function fetchInventoryDefaults(company?: string): Promise<InventoryDefaults> {
  return apiRequest<InventoryDefaults>(`/api/method/${API}.get_inventory_defaults`, {
    method: 'POST',
    body: JSON.stringify({ company: company || null }),
  });
}

export async function fetchStockBalanceReport(params: {
  company?: string;
  warehouse?: string;
  item_code?: string;
  item_group?: string;
  search?: string;
  as_on_date?: string;
  sort_order?: 'asc' | 'desc';
  limit?: number;
}): Promise<StockBalanceReport> {
  return apiRequest<StockBalanceReport>(`/api/method/${API}.get_stock_balance_report`, {
    method: 'POST',
    body: JSON.stringify(params),
  });
}

export async function fetchStockLedgerReport(params: {
  company?: string;
  warehouse?: string;
  item_code?: string;
  item_group?: string;
  search?: string;
  from_date?: string;
  to_date?: string;
  limit?: number;
}): Promise<StockLedgerReport> {
  return apiRequest<StockLedgerReport>(`/api/method/${API}.get_stock_ledger_report`, {
    method: 'POST',
    body: JSON.stringify(params),
  });
}

export async function fetchInventoryInsights(params: {
  company?: string;
  warehouse?: string;
  from_date?: string;
  to_date?: string;
  low_stock_limit?: number;
  consumed_limit?: number;
}): Promise<InventoryInsightsReport> {
  return apiRequest<InventoryInsightsReport>(`/api/method/${API}.get_inventory_insights_report`, {
    method: 'POST',
    body: JSON.stringify(params),
  });
}
